consensus/core/src/leader_scoring.rs

```diff
@@ -138,173 +69,214 @@ impl ReputationScores {
+    ) {
+        let _s = self
+            .context
+            .metrics
+            .node_metrics
+            .scope_processing_time
+            .with_label_values(&["ScoringSubdag::add_unscored_committed_subdags"])
+            .start_timer();
+        for subdag in committed_subdags {
+            // If the commit range is not set, then set it to the range of the first
+            // committed subdag index.
+            if self.commit_range.is_none() {
+                self.commit_range = Some(CommitRange::new(
+                    subdag.commit_ref.index..=subdag.commit_ref.index,
+                ));
+            } else {
+                let commit_range = self.commit_range.as_mut().unwrap();
+                commit_range.extend(subdag.commit_ref.index);
             }
-            // Weak links may point to blocks with lower round numbers than strong links.
-            if ancestor.round <= leader_slot.round {
-                continue;
+
+            // Add the committed leader to the list of leaders we will be scoring.
+            tracing::trace!("Adding new committed leader {} for scoring", subdag.leader);
+            self.leaders.insert(subdag.leader);
+
+            // Check each block in subdag. Blocks are in order so we should traverse the
+            // oldest blocks first
+            for block in subdag.blocks {
+                for ancestor in block.ancestors() {
+                    // Weak links may point to blocks with lower round numbers
+                    // than strong links.
+                    if ancestor.round != block.round().saturating_sub(1) {
+                        continue;
+                    }
+
+                    // If a blocks strong linked ancestor is in leaders, then
+                    // it's a vote for leader.
+                    if self.leaders.contains(ancestor) {
+                        // There should never be duplicate references to blocks
+                        // with strong linked ancestors to leader.
+                        tracing::trace!(
+                            "Found a vote {} for leader {ancestor} from authority {}",
+                            block.reference(),
+                            block.author()
+                        );
+                        assert!(self
+                            .votes
+                            .insert(block.reference(), StakeAggregator::new())
+                            .is_none(), "Vote {block} already exists. Duplicate vote found for leader {ancestor}");
+                    }
+
+                    if let Some(stake) = self.votes.get_mut(ancestor) {
+                        // Vote is strongly linked to a future block, so we
+                        // consider this a certified vote.
+                        tracing::trace!(
+                            "Found a certified vote {ancestor} from authority {}",
+                            ancestor.author
+                        );
```

Comment:
I don't think the comment and log is accurate here? This vote by ancestor may not be certified until there is enough stake.